File: getting_started/234.回文链表.py
#
# @lc app=leetcode.cn id=234 lang=python3
#
# [234] 回文链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        # brute force
        nodes = []
        while head is not None:
            nodes.append(head.val)
            head = head.next
        r = len(nodes) - 1
        l = 0
        while l < r:
            if nodes[l] != nodes[r]:
                return False
            l += 1
            r -= 1
        return True
# @lc code=end

File: getting_started/234.回文链表.cpp
/*
 * @lc app=leetcode.cn id=234 lang=cpp
 *
 * [234] 回文链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        std::vector<int> vals;
        while(head)
        {
            vals.push_back(head->val);
            head = head->next;
        }
        int left = 0;
        int right = vals.size() - 1;
        while(left < right)
        {
            if (vals[left] != vals[right])
            {
                return false;
            }
            left ++;
            right --;
        }
        return true;
    }
};
// @lc code=end

